// unplugin-vue-define-options/dist/chunk-EUTA4OJI.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _chunkLZS2JEKLjs = require('./chunk-LZS2JEKL.js');

// src/unplugin.ts
var _unplugin = require('unplugin');
var _pluginutils = require('@rollup/pluginutils');




var _common = require('@vue-macros/common');
function resolveOption(options) {
  return {
    include: options.include || [/\.vue$/],
    exclude: options.exclude || void 0
  };
}
var unplugin_default = _unplugin.createUnplugin.call(void 0, (options = {}) => {
  const opt = resolveOption(options);
  const filter = _pluginutils.createFilter.call(void 0, opt.include, opt.exclude);
  const name = "unplugin-vue-define-options";
  return {
    name,
    enforce: "pre",
    transformInclude(id) {
      return filter(id);
    },
    transform(code, id) {
      try {
        const { ctx, getMagicString } = _common.initContext.call(void 0, code, id);
        _chunkLZS2JEKLjs.transform.call(void 0, ctx);
        _common.finalizeContext.call(void 0, ctx);
        return _common.getTransformResult.call(void 0, getMagicString(), id);
      } catch (err) {
        this.error(`${name} ${err}`);
      }
    }
  };
});



exports.unplugin_default = unplugin_default;
