// @vue-macros/common/dist/chunk-NEMJ554U.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }


var _chunkEYOCX5COjs = require('./chunk-EYOCX5CO.js');

// src/vue/context.ts
var _magicstring = require('magic-string'); var _magicstring2 = _interopRequireDefault(_magicstring);
var initContext = (code, id) => {
  let s;
  let sfc;
  const ctx = {
    code,
    id,
    scriptCode: {
      prepend: "",
      append: ""
    },
    get s() {
      return s || (s = new (0, _magicstring2.default)(code));
    },
    set s(_s) {
      s = _s;
    },
    get sfc() {
      return sfc || (sfc = _chunkEYOCX5COjs.parseSFC.call(void 0, code, id));
    },
    set sfc(_sfc) {
      sfc = _sfc;
    }
  };
  return {
    ctx,
    getMagicString() {
      return s;
    },
    getSFC() {
      return sfc;
    }
  };
};
var finalizeContext = (ctx) => {
  _chunkEYOCX5COjs.addToScript.call(void 0, ctx);
};




exports.initContext = initContext; exports.finalizeContext = finalizeContext;
